### ISSUE TITLE
Support for LSTM/GRUs

### ISSUE THREAD

**Issue by ak-7** [musically-ut/tf_rmtpp#7]
Hello,

Are you planning to extend support to LSTM/GRUs?
From your code I see all the RNN functionality is not encapsulated in any RNN class. Currently, hidden states creation and updation code is intertwined with other variables. Do you have something in mind for easiest way to extend it to LSTMs + GRUs?

**Comment by musically-ut** [comment]
Such an extension is not in the pipeline.

The functionality is not embedded inside an RNNCell because truncated BPTT, which was needed for longer sequences, is much trickier to implement in that setting. You can see examples of such cells in [TPPRL](https://github.com/Networks-Learning/tpprl/blob/master/tpprl/cells.py), where we did not employ BPTT but instead kept the episode lengths controlled.

The extension to LSTMs + GRUs is not too difficult, though, because the only thing which would change is the formula for calculating `h[i+1]` from `h[i]`, i.e. Eq. (9) in the [paper](https://www.kdd.org/kdd2016/papers/files/rpp1081-duA.pdf).

**Comment by musically-ut** [comment]
Please close the issue if the question was answered. 

Thanks!

**Comment by ak-7** [comment]
Thanks!

Could you elaborate a bit on why BPTT is trickier to implement with RNNCell? My understanding is in a batch if you've reached the end of a sequence, the events will be zero and we can ignore those updates. What am I missing?

**Comment by musically-ut** [comment]
`RNNCell` can indeed handle end of sequences and, hence, can work with batches with sequences of different sizes. However, truncated-BPTT is about limiting the number of elements over which the gradient is calculated, which is a very different problem, requires blocking the _flow_ of gradients after `N` steps (where `N` is the truncation horizon for BPTT). This can be done using ['tf.stop_gradients`](https://www.tensorflow.org/api_docs/python/tf/stop_gradient) but it requires injecting extra information into the Cell state and is messy in general. 

I hope that makes the problem with BPTT clearer.

That said, I haven't looked around to see if there are any good ways of doing BPTT while using RNNCell with the recent versions of TensorFlow (or PyTorch). If you find such an example, please do share it with me and I would be happy to take a look.

**Comment by ak-7** [comment]
Alright. Thanks for the explanation.
